### PR TITLE
Rework store_inplace

### PR DESCRIPTION
**Pull request by bmerry**
A lot of the code in `store_inplace` pre-dates HighLevelGraph in Dask, and tended to materialize low-level graphs earlier than necessary and thus potentially disable high-level optimisations. It's rewritten using high-level layers, so that other than the safety check and the actual computation, performance should only depend on the number of layers involved rather than the number of chunks.

It also fixes some of the `rename` code to work with newer Dask versions that have extra fields in Blockwise layers.

### CODE REVIEW

**Comment by bmerry** [comment]
Note to myself (to check on next week): I have a feeling that `rename` might be a re-implementation of `dask.graph_manipulation.clone`, in which case it could be eliminated.

**Comment by bmerry** [comment]
> Note to myself (to check on next week): I have a feeling that rename might be a re-implementation of dask.graph_manipulation.clone, in which case it could be eliminated.

dask.graph_manipulation.clone does appear to be exactly what I need, but unfortunately is not available in the Dask version we've pinned to. So that simplification will need to wait until https://github.com/dask/dask/pull/7637 lands in Dask and we can upgrade to a newer version.

File: katsdpcal/test/test_inplace.py [hunk 1]
```diff
@@ -3,19 +3,28 @@
...
 import numpy as np
 import dask.array as da
+import dask.distributed
 
 from katsdpcal import inplace
 
 
 class TestStoreInplace:
     def setup(self):
+        self.cluster = dask.distributed.LocalCluster(
```
Comment (ludwigschwardt) [inline katsdpcal/test/test_inplace.py:13]: What does a local cluster gain you?

File: katsdpcal/inplace.py [hunk 2]
```diff
@@ -187,31 +215,55 @@ def store(target, source):
...
+    # we'll gather up all the output keys into one "root" key and form a
+    # Delayed collection from it. This is similar to what da.store does.
+    root_key = 'store-root-' + str(uuid.uuid4())
+    layers[root_key] = {root_key: out_keys}
+    dependencies[root_key] = set(store_layers)
+    graph = HighLevelGraph(layers, dependencies)
+    # Ensure that array-appropriate optimizations are performed.
+    graph = da.Array.__dask_optimize__(graph, [root_key])
+    result = Delayed(root_key, graph)
+    result.compute(optimize_graph=False)
```
Comment (ludwigschwardt) [inline katsdpcal/inplace.py:266]: Is the parameter because of the earlier `__dask_optimize__`?

File: katsdpcal/inplace.py [hunk 3]
```diff
@@ -108,7 +139,7 @@ def _safe_in_place(dsk, source_keys, target_keys):
             arrays[k] = out
     for key in target_keys:
         if len(arrays[key]) != 1 or not next(iter(arrays[key])).elementwise:
-            raise ValueError('Target key {} does not directly map a numpy array')
+            raise ValueError(f'Target key {key} does not directly map a numpy array')
```
Comment (ludwigschwardt) [inline katsdpcal/inplace.py:142]: Nice :-)

File: katsdpcal/inplace.py [hunk 4]
```diff
@@ -65,7 +78,19 @@ def _in_graph(dsk, key):
...
+    # more complicated cases where the parameters are not simply a
+    # node and a literal expression.
+    return (type(v) is tuple and len(v) == 3
+            and v[0] in da.optimization.GETTERS
+            and _in_graph(dsk, v[1])
+            and not dask.core.has_tasks(dsk, v[2]))
+
+
+def _safe_inplace(sources, targets):
     """Safety check on :func:`safe_in_place`. It uses the following algorithm:
```
Comment (ludwigschwardt) [inline katsdpcal/inplace.py:94]: Should this be `store_inplace` instead?

**Review by ludwigschwardt — COMMENTED** [review]
I have a very tenuous grasp on the correctness, but overall looks good.

**Review by ludwigschwardt — APPROVED** [review]

**Comment by bmerry** [comment]
> I have a very tenuous grasp on the correctness

Thanks. I've asked @JSKenyon if he wouldn't mind having a look too.

**Comment by JSKenyon** [comment]
Looks good to me, though I would need to understand the problem that is being solved a bit better to be more helpful.

**Comment by bmerry** [comment]
> Looks good to me, though I would need to understand the problem that is being solved a bit better to be more helpful.

Thanks. Unfortunately Friday lightning talks at the summit got cancelled, otherwise I would have explained there. The short, short version is that we use up most of the memory in the machine to hold visibilities so there isn't enough memory to do a computation that writes its outputs to a whole new array.

**Pull request merged**